File: frontends/android/app/src/test/java/com/example/animastor/repository/ResilientReloaderTest.kt
package com.example.animastor.repository

import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.async
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.withTimeout
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import java.io.IOException

/**
 * Reload retry/recovery state machine (ResilientReloader) — the layer on top
 * of the ResourceInvalidations pipeline:
 *
 *     invalidation → reload → network failure → retry/recovery → reload
 *
 * Pure JVM with a fake [NetworkRecoverySignal]; zero-length backoff steps keep
 * the tests fast while preserving the attempt/phase logic.
 */
class ResilientReloaderTest {

    /** Fake connectivity signal: manual online/offline flips with a real
     *  SharedFlow for restore events. */
    private class FakeRecovery(startOnline: Boolean = true) : NetworkRecoverySignal {
        private val flow = MutableSharedFlow<Unit>(replay = 0, extraBufferCapacity = 8)
        override val restored = flow.asSharedFlow()
        @Volatile override var isOnline: Boolean = startOnline
        fun goOffline() { isOnline = false }
        fun goOnline() {
            isOnline = true
            flow.tryEmit(Unit)
        }
    }

    private val noBackoff = listOf(0L, 0L, 0L)

    @Test
    fun `success on first attempt - no retries`() = runBlocking {
        val recovery = FakeRecovery()
        val reloader = ResilientReloader(recovery, noBackoff)
        var calls = 0
        val result = reloader.reload {
            calls++
            "fresh"
        }
        assertTrue(result is ResilientReloader.Result.Success)
        assertEquals("fresh", (result as ResilientReloader.Result.Success).value)
        assertEquals(1, calls)
    }

    @Test
    fun `transient failure is retried until success`() = runBlocking {
        val recovery = FakeRecovery()
        val reloader = ResilientReloader(recovery, noBackoff)
        var calls = 0
        val result = reloader.reload {
            calls++
            if (calls == 1) throw IOException("offline blip")
            "fresh"
        }
        assertTrue(result is ResilientReloader.Result.Success)
        assertEquals("fresh", (result as ResilientReloader.Result.Success).value)
        assertEquals(2, calls)
    }

    @Test
    fun `gives up after the bounded cycle when network stays down (online)`() = runBlocking {
        val recovery = FakeRecovery() // online but every request fails (server-side timeouts)
        val reloader = ResilientReloader(recovery, noBackoff)
        var calls = 0
        val result = reloader.reload {
            calls++
            throw IOException("timeout")
        }
        assertTrue(result is ResilientReloader.Result.TransientFailure)
        assertEquals(reloader.phaseOneAttempts, calls)
        assertEquals(reloader.phaseOneAttempts, (result as ResilientReloader.Result.TransientFailure).attempts)
    }

    @Test
    fun `permanent failure (server answered) is never retried`() = runBlocking {
        val recovery = FakeRecovery()
        val reloader = ResilientReloader(recovery, noBackoff)
        var calls = 0
        val result = reloader.reload {
            calls++
            throw RuntimeException("HTTP 404") // non-IOException = server answered
        }
        assertTrue(result is ResilientReloader.Result.PermanentFailure)
        assertEquals(1, calls)
    }

    @Test
    fun `parks offline and makes the final attempt when connectivity returns`() = runBlocking {
        val recovery = FakeRecovery(startOnline = false)
        val reloader = ResilientReloader(recovery, noBackoff)
        var calls = 0
        val waiter = async {
            withTimeout(5000) {
                reloader.reload {
                    calls++
                    // Succeed only on the final post-restore attempt.
                    if (calls <= reloader.phaseOneAttempts) throw IOException("still offline")
                    "recovered"
                }
            }
        }
        // Let the bounded cycle exhaust while offline, then restore.
        delay(100)
        recovery.goOnline()
        val result = waiter.await()
        assertTrue(result is ResilientReloader.Result.Success)
        assertEquals("recovered", (result as ResilientReloader.Result.Success).value)
        assertEquals(reloader.phaseOneAttempts + 1, calls)
    }

    @Test
    fun `restore during backoff short-circuits the wait (fast path)`() = runBlocking {
        val recovery = FakeRecovery()
        val reloader = ResilientReloader(
            recovery,
            retryDelaysMs = listOf(60_000L, 60_000L, 60_000L, 60_000L) // would take minutes
        )
        var calls = 0
        val waiter = async {
            withTimeout(5000) {
                reloader.reload {
                    calls++
                    if (calls == 1) throw IOException("blip right before restore")
                    "fast"
                }
            }
        }
        // The restore event must release the 60s backoff wait almost immediately.
        delay(50)
        recovery.goOnline()
        val result = waiter.await()
        assertTrue(result is ResilientReloader.Result.Success)
        assertEquals(2, calls)
    }

    @Test
    fun `onRetry hook fires on every scheduled backoff`() = runBlocking {
        val recovery = FakeRecovery()
        val reloader = ResilientReloader(recovery, noBackoff)
        val retryAttempts = mutableListOf<Int>()
        var calls = 0
        reloader.reload(
            onRetry = { attempt, _ -> retryAttempts.add(attempt) }
        ) {
            calls++
            if (calls <= 2) throw IOException("flaky")
            Unit
        }
        assertEquals(listOf(1, 2), retryAttempts)
    }

    @Test
    fun `cancellation propagates - never swallowed as a failure`() = runBlocking {
        val recovery = FakeRecovery()
        val reloader = ResilientReloader(recovery, noBackoff)
        val gate = CompletableDeferred<Unit>()
        val job = launch {
            try {
                reloader.reload {
                    gate.await() // suspended "request"
                    throw IOException("never observed")
                }
            } catch (e: CancellationException) {
                gate.complete(Unit)
                throw e
            }
        }
        delay(50)
        job.cancel()
        job.join()
        assertTrue(job.isCancelled)
    }
}

File: frontends/app/src/state/resilientReloader.test.ts
// Tests for the reload retry/recovery state machine (resilientReloader.ts) —
// the web half of the layer on top of the ResourceInvalidations pipeline.
// Android parity: ResilientReloaderTest.kt. Fake recovery signal; zero-length
// backoff steps keep the tests fast while preserving attempt/phase logic.
import { describe, it, expect } from 'vitest';
import {
  resilientReload, isTransientError, phaseOneAttempts,
  DEFAULT_RETRY_DELAYS_MS, type NetworkRecoverySignal,
} from './resilientReloader';
import { ApiError } from '../api/client';

/** Fake connectivity signal with manual online/offline flips. */
class FakeRecovery implements NetworkRecoverySignal {
  online: boolean;
  private cbs = new Set<() => void>();
  constructor(startOnline = true) {
    this.online = startOnline;
  }
  get isOnline(): boolean {
    return this.online;
  }
  onNextRestore(fn: () => void): () => void {
    this.cbs.add(fn);
    return () => {
      this.cbs.delete(fn);
    };
  }
  goOffline(): void {
    this.online = false;
  }
  goOnline(): void {
    this.online = true;
    this.cbs.forEach((fn) => fn());
    this.cbs.clear();
  }
}

const noBackoff = [0, 0, 0];

describe('resilientReload', () => {
  it('succeeds on first attempt without retries', async () => {
    const recovery = new FakeRecovery();
    let calls = 0;
    const result = await resilientReload({
      recovery,
      attempt: async () => {
        calls++;
        return 'fresh';
      },
      retryDelaysMs: noBackoff,
    });
    expect(result).toEqual({ kind: 'success', value: 'fresh' });
    expect(calls).toBe(1);
  });

  it('retries a transient failure until success', async () => {
    const recovery = new FakeRecovery();
    let calls = 0;
    const result = await resilientReload({
      recovery,
      attempt: async () => {
        calls++;
        if (calls === 1) throw new TypeError('Failed to fetch'); // offline blip
        return 'fresh';
      },
      retryDelaysMs: noBackoff,
    });
    expect(result).toEqual({ kind: 'success', value: 'fresh' });
    expect(calls).toBe(2);
  });

  it('gives up after the bounded cycle when the server keeps timing out (online)', async () => {
    const recovery = new FakeRecovery(); // online but every request times out
    let calls = 0;
    const result = await resilientReload({
      recovery,
      attempt: async () => {
        calls++;
        throw new ApiError('Request timeout', 408);
      },
      retryDelaysMs: noBackoff,
    });
    expect(result.kind).toBe('transient-failure');
    expect(calls).toBe(phaseOneAttempts(noBackoff));
  });

  it('permanent failure (server answered) is never retried', async () => {
    const recovery = new FakeRecovery();
    let calls = 0;
    const result = await resilientReload({
      recovery,
      attempt: async () => {
        calls++;
        throw new ApiError('Not Found', 404); // the server IS reachable
      },
      retryDelaysMs: noBackoff,
    });
    expect(result).toMatchObject({ kind: 'permanent-failure' });
    expect(calls).toBe(1);
  });

  it('parks offline and makes the final attempt when connectivity returns', async () => {
    const recovery = new FakeRecovery(false);
    let calls = 0;
    const p = resilientReload({
      recovery,
      attempt: async () => {
        calls++;
        if (calls <= phaseOneAttempts(noBackoff)) throw new TypeError('still offline');
        return 'recovered';
      },
      retryDelaysMs: noBackoff,
    });
    // Let the bounded cycle exhaust while offline, then restore.
    await new Promise((r) => setTimeout(r, 20));
    recovery.goOnline();
    const result = await p;
    expect(result).toEqual({ kind: 'success', value: 'recovered' });
    expect(calls).toBe(phaseOneAttempts(noBackoff) + 1);
  });

  it('restore during backoff short-circuits the wait (fast path)', async () => {
    const recovery = new FakeRecovery();
    let calls = 0;
    const p = resilientReload({
      recovery,
      attempt: async () => {
        calls++;
        if (calls === 1) throw new TypeError('blip right before restore');
        return 'fast';
      },
      retryDelaysMs: [60_000, 60_000, 60_000, 60_000], // would take minutes
    });
    // The restore event must release the 60s backoff wait almost immediately.
    await new Promise((r) => setTimeout(r, 10));
    recovery.goOnline();
    const result = await p;
    expect(result).toEqual({ kind: 'success', value: 'fast' });
    expect(calls).toBe(2);
  });

  it('onRetry hook fires on every scheduled backoff', async () => {
    const recovery = new FakeRecovery();
    const retryAttempts: number[] = [];
    let calls = 0;
    await resilientReload({
      recovery,
      attempt: async () => {
        calls++;
        if (calls <= 2) throw new TypeError('flaky');
        return null;
      },
      retryDelaysMs: noBackoff,
      onRetry: (attempt) => retryAttempts.push(attempt),
    });
    expect(retryAttempts).toEqual([1, 2]);
  });

  it('timeout ApiError (408) is transient, other statuses are not', () => {
    expect(isTransientError(new ApiError('Request timeout', 408))).toBe(true);
    expect(isTransientError(new ApiError('Not Found', 404))).toBe(false);
    expect(isTransientError(new ApiError('Server Error', 500))).toBe(false);
    expect(isTransientError(new TypeError('Failed to fetch'))).toBe(true);
  });

  it('default backoff steps match the Android contract (1s 2s 5s 10s)', () => {
    expect(DEFAULT_RETRY_DELAYS_MS).toEqual([1_000, 2_000, 5_000, 10_000]);
    expect(phaseOneAttempts(DEFAULT_RETRY_DELAYS_MS)).toBe(5);
  });
});
